### PR TITLE
[CALCITE-4512] GROUP BY expression with argument name same with SELECT field and alias causes validation error

### PR DESCRIPTION
**Pull request by hannerwang**
### **What changes were proposed in this pull request?**
Stop identifiers in a SQL call from being treated as aliases and expanded into select expressions.
For example:
`select replace(name, 'a', 'b') as name from users group by replace(name, 'a', 'b')`
The **name** in GROUP BY clause should not be replaced by `replace(name, 'a', 'b')`, it will be resolved as users.name,
i.e. 
`select replace(users.name, 'a', 'b') as name from users group by replace(users.name, 'a', 'b')`

### **Why are the changes needed?**
Make the sql semantics unchanged when treating identifiers within Group By clause as aliases.
current calcite behavior:
`select replace(**name**, 'a', 'b') as **name** from users group by replace(**name**, 'a', 'b')`
will be expanded into
`select replace(**name**, 'a', 'b') as **name** from users group by replace(**replace(name, 'a', 'b')**, 'a', 'b')`
the semantics has been changed.

### **Does this PR introduce any user-facing change?**
No for existing users

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java
Comment (suibianwanwank) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java:7385]: Should SqlBasicCall be replaced with SqlCall here. e.g. SqlCase does not inherit SqlBasicCall.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java
Comment (suibianwanwank) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java:7386]: Same as above.

**Comment by mihaibudiu** [comment]
@hannerwang I don't know if you have seen the comments in the JIRA.
This PR is very close to being done.
Do you think you can address the comments in time for the next release of Calcite?

I reproduce the comments here:

> Which DB are we trying to emulate the behavior of? If so, I would like to see a Quidem test that returns the same in Calcite and that DB, and would return the wrong result if the bug were not fixed. Perhaps something like select mod(empno, 3) as empno from emp group by mod(empno, 2) and make sure that we get the right number of rows.
>
> Also I would like to see a SqlValidatorTest. Something like select date_add('2024-01-01', empno) as empno from emp group by year(empno), where there will be a type error (int vs date) if the wrong column is referenced.
> 
> In short, more tests - that are broken before the fix, and pass after the fix, and are demonstrably the same behavior as a given DB - will give me the confidence to approve this PR.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java
Comment (hannerwang) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java:7385]: @suibianwanwank Thanks your remind, the switch check ensures the root is a basic call here.

**Comment by hannerwang** [comment]
@mihaibudiu thanks, I have done more tests and add new comment. I'm available to push the pr merged this week.

**Comment by mihaibudiu** [comment]
For now maybe you can make the CI pass

**Comment by hannerwang** [comment]
> For now maybe you can make the CI pass

Hello @mihaibudiu , do you know how to rerun the CI/macos manually? the CI fails unexpectedly.

**Comment by mihaibudiu** [comment]
This is how you can rerun failed jobs:
image
if you have enough privileges

**Comment by hannerwang** [comment]
> This is how you can rerun failed jobs: image if you have enough privileges

Thanks, I think I have no the privilege to rerun.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java:7390]: Can't ROWS contain other ROWS as fields?
In that case do you need a recursive call?

File: core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java
Comment (mihaibudiu) [inline core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java:6317]: I don't really understand this test.

Comment (hannerwang) [reply core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java:6317]: now date_add function receives empno:integer as a parameter, returns a DATE type.
if empno in year(empno) is substituted by 'date_add('2024-01-01', empno)', the year function will not raise type inconsistency error.
but we expect it throws here, because the empno in year(empno) should not be substituted. the identifier in a function should not be substituted.
I add this test as @julianhyde has mentioned this case. I paste his comment here.

> Also I would like to see a SqlValidatorTest. Something like select date_add('2024-01-01', empno) as empno from emp group by year(empno), where there will be a type error (int vs date) if the wrong column is referenced.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java
Comment (hannerwang) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: Thanks @mihaibudiu , I test on bigquery and spark, they behave as you mentioned, we should consider the recursive rows, I added a method to check the nested rows case, you can review it when you are available.

**Review by mihaibudiu — APPROVED** [review]

**Comment by mihaibudiu** [comment]
This PR is waiting for @julianhyde's approval, based on the discussion in JIRA.

**Pull request merged**
